underscore.js

```diff
@@ -1109,12 +1109,19 @@
 
 
   // Internal recursive comparison function for `isEqual`.
-  var eq = function(a, b, aStack, bStack) {
-    // Identical objects are equal. `0 === -0`, but they aren't identical.
-    // See the [Harmony `egal` proposal](http://wiki.ecmascript.org/doku.php?id=harmony:egal).
-    if (a === b) return a !== 0 || 1 / a === 1 / b;
-    // A strict comparison is necessary because `null == undefined`.
-    if (a == null || b == null) return a === b;
+  var deepEq, eq = function(a, b, aStack, bStack) {
+      // Identical objects are equal. `0 === -0`, but they aren't identical.
+      // See the [Harmony `egal` proposal](http://wiki.ecmascript.org/doku.php?id=harmony:egal).
+      if (a === b) return a !== 0 || 1 / a === 1 / b;
+
+      // if objects are of primitive types, they are not equal unless both of them are NaNs.
+      if (typeof a != 'object' && typeof b != 'object') return a != a && b != b;
+
+      // if any of a or b is null, objects are not equal.
+      return a != null && b != null && deepEq(a, b, aStack, bStack);
+  };
+
+  deepEq = function(a, b, aStack, bStack) {
```

Comment:
eq is recursive function, so all the logic we create must be recursive too. Also, splitting this function into two dramatically improved performance in Firefox JIT. It was slower than lodash before this change.